### DISCUSSION TITLE
Delay initialization till usage

### DISCUSSION THREAD

**Discussion by NN---** [dotnet/csharplang#3563] [Language Ideas]
It could be a nice optimization if there was possibility to delay initialization of variables and field till their first usage.

Today you can make lazy initialization using System.Lazy or property with getter.
But this means that variable is created on access and not on the actual usage.

Sample:

```cs
class A
{
public delay init List<int> L = new List<int>();
}

a = new A();
var x = a.L; // L is still null here
a.L.Add(1); // here L is initialized to List<int>
```

How this improves performance:
In many cases there are classes with default collection initialization meaning we create a lot of unnecessary objects which can be prevented.

**Comment by CyrusNajmabadi** [comment]
`var x = a.L; // L is still null here`

THat seems super not good.  THat's a read of the property.  Having that return null means people could trivially crash when trying to read props and use the values returned by them.

**Comment by HaloFour** [comment]
This is what [`Lazy<T>`](https://docs.microsoft.com/en-us/dotnet/api/system.lazy-1) is for.  There's no other way in the runtime to have a reference type be "lazily initialized".  The reference has to either point to an instance of the type, or to `null`.

**Comment by HaloFour** [comment]
This is a dupe of your own closed #3539.

**Comment by NN---** [comment]
The problem is that Lazy is a class and it means allocation on heap.

**Comment by NN---** [comment]
> The problem is that Lazy is a class and it means allocation on heap.

Yes ,but as it is mentioned there probably the best way is to do it from runtime perspective.
This why I opened it here.

**Comment by HaloFour** [comment]
@NN--- 

> The problem is that Lazy is a class and it means allocation on heap.

Yes it does.  You need something that:

1. Tracks initialization
2. Can be shared between multiple locations
3. Isn't tied to the stack of the method/property that is returning it

In other words, you have no choice but to have this abstraction live on the heap.  Doesn't matter if you, the language or the runtime does it.

**Comment by CyrusNajmabadi** [comment]
> Yes ,but as it is mentioned there probably the best way is to do it from runtime perspective.

This is not the runtime though :)   This is the C# language.

**Comment by CyrusNajmabadi** [comment]
> The problem is that Lazy is a class and it means allocation on heap.

You can use LazyInitializer: https://docs.microsoft.com/en-us/dotnet/api/system.threading.lazyinitializer?view=netcore-3.1

**Comment by CyrusNajmabadi** [comment]
In order for thsi to be taken seriously as a language feature, it will need to come with an actual explanation of how this would work.  i.e. given your code:

```c#
class A
{
   public delay init List<int> L = new List<int>();
}

a = new A();
var x = a.L; // L is still null here
a.L.Add(1); // here L is initialized to List<int>
```

What does that compile down to.  Please ensure you consider things like teh cross-project case where the class/field are declared in one project but referenced in another.

**Comment by NN---** [comment]
Agree
In order to do this it may require runtime change and language support.

**Comment by NN---** [comment]
> > Yes ,but as it is mentioned there probably the best way is to do it from runtime perspective.
> 
> This is not the runtime though :) This is the C# language.

It is also runtime since it needs to know to the usage somehow .

**Comment by CyrusNajmabadi** [comment]
> It is also runtime since it needs to know to the usage somehow .

I cannot see the runtime taking on a feature to do this when it's already simple to just use things like LazyInitializer.

Can you also clarify how the runtime portion woudl actually work?  You say, for example, "since it needs to know to the usage somehow".  But how are you seeing that happening without an allocation of some sort?

**Comment by DaZombieKiller** [comment]
With https://github.com/dotnet/csharplang/issues/140 you can do this:
```CSharp
public List<int> L => LazyInitializer.EnsureInitialized(ref field);
```

**Comment by NN---** [comment]
I would like to be able to have no GC overhead when the field is not used.

**Comment by DaZombieKiller** [comment]
The example above would give you that exact behaviour, as the field would only be initialized upon the first usage.
